src/widgets/docsViewer.js

```diff
@@ -109,14 +109,25 @@ var FlatsealDocsViewer = GObject.registerClass({
         const uri = action.get_request().get_uri();
 
         if (!uri.startsWith('file')) {
-            Gio.AppInfo.launch_default_for_uri(uri, null);
+            const launcher = new Gtk.UriLauncher();
+            launcher.uri = uri;
+            launcher.launch(this, null, this._OpenUriCb);
+
             decision.ignore();
             return true;
         }
 
         return false;
     }
 
+    _OpenUriCb(launcher, res) {
+        try {
+            launcher.launch_finish(res);
+        } catch(err) {
+            logError(err);
+        }
+    }
+
```

Comment:
To make this PR simpler to review and merge, I suggest to leave misc improvements out. Let's just focus strictly on migrating to GTK 4 and Libadwaita.